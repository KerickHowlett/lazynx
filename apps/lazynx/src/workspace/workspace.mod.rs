mod workspace_store;

mod workspace_view_widget;
pub use workspace_view_widget::WorkspaceViewWidget;

mod workspace_tab_widget;
pub use workspace_tab_widget::WorkspaceTabWidget;

#[cfg(test)]
#[path = "./test_utils/workspace_test_utils.mod.rs"]
pub mod test_bed;
