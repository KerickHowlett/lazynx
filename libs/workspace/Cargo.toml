[package]
name = "workspace"
version = "0.1.0"
edition = "2021"

[dependencies]
# Local Dependencies
app_config = { path = "../../libs/shared/config" }

# Third-Party Dependencies
chrono = "0.4.38"
color-eyre = "0.6.3"
crossterm = { version = "0.28.1", features = ["serde", "event-stream"] }
ratatui = { version = "0.29.0", features = ["serde", "macros"] }

[dev-dependencies]
insta = "1.42.0"
