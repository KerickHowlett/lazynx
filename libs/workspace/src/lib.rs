mod workspace_view_widget;
pub use workspace_view_widget::WorkspaceViewWidget;

mod workspace_tab_widget;
pub use workspace_tab_widget::WorkspaceTabWidget;
