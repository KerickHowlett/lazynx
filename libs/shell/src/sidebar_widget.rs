use std::rc::Rc;

use ratatui::prelude::{Constraint, Direction, Frame, Layout, Rect};

use workspace::WorkspaceTabWidget;

#[derive(Default)]
pub struct SidebarWidget {
    workspace_tab: WorkspaceTabWidget,
}

impl SidebarWidget {
    pub fn draw(&mut self, frame: &mut Frame, area: Rect) {
        let chunks = self.create_layout(area);

        self.workspace_tab.draw(frame, chunks[0]);
    }

    pub fn init(&mut self) -> color_eyre::eyre::Result<()> {
        self.workspace_tab.init()?;
        Ok(())
    }

    fn create_layout(&self, area: Rect) -> Rc<[Rect]> {
        return Layout::default()
            .direction(Direction::Vertical)
            .constraints(vec![Constraint::Length(3), Constraint::Fill(1)])
            .split(area);
    }
}

#[cfg(test)]
mod sidebar_widget_tests {
    use super::SidebarWidget;

    use insta::assert_snapshot;

    use test_utils::WidgetTestBed;

    use workspace::test_bed::WorkspaceTestBed;

    struct TestBed {
        widget: WidgetTestBed<SidebarWidget>,
        workspace: WorkspaceTestBed,
    }

    impl Default for TestBed {
        fn default() -> Self {
            return TestBed {
                workspace: WorkspaceTestBed::default(),
                widget: WidgetTestBed::<SidebarWidget>::new(30, 50),
            };
        }
    }

    #[test]
    fn test_sidebar_widget_draw() {
        let mut test_bed = TestBed::default();
        test_bed.workspace.setup();

        test_bed.widget.widget.init().unwrap();
        test_bed
            .widget
            .terminal
            .draw(|f| test_bed.widget.widget.draw(f, f.area()))
            .unwrap();

        assert_snapshot!(test_bed.widget.terminal.backend());

        test_bed.workspace.restore();
    }
}
