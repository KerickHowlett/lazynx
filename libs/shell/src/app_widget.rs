use ratatui::{
    layout::Rect,
    prelude::{Constraint, Direction, Layout},
    Frame,
};

use color_eyre::eyre::Result;
use workspace::WorkspaceViewWidget;

use crate::sidebar_widget::SidebarWidget;

pub trait IAppWidget {
    fn draw(&mut self, frame: &mut Frame, area: Rect);
    fn init(&mut self) -> Result<()>;
}

#[derive(Default)]
pub struct AppWidget {
    sidebar: SidebarWidget,
    workspace: WorkspaceViewWidget,
}

impl IAppWidget for AppWidget {
    fn draw(&mut self, frame: &mut Frame, _area: Rect) {
        let chunks = Layout::default()
            .direction(Direction::Horizontal)
            .constraints(vec![Constraint::Min(38), Constraint::Percentage(75)])
            .split(frame.area());

        self.sidebar.draw(frame, chunks[0]);
        self.workspace.draw(frame, chunks[1]);
    }

    fn init(&mut self) -> color_eyre::eyre::Result<()> {
        self.sidebar.init()?;
        Ok(())
    }
}

#[cfg(test)]
mod app_widget_tests {
    use super::{AppWidget, IAppWidget};

    use insta::assert_snapshot;

    use test_utils::WidgetTestBed;
    use workspace::test_bed::WorkspaceTestBed;

    struct TestBed {
        widget: WidgetTestBed<AppWidget>,
        workspace: WorkspaceTestBed,
    }

    impl Default for TestBed {
        fn default() -> Self {
            return TestBed {
                workspace: WorkspaceTestBed::default(),
                widget: WidgetTestBed::<AppWidget>::new(30, 50),
            };
        }
    }

    #[test]
    fn test_sidebar_widget_draw() {
        let mut test_bed = TestBed::default();
        test_bed.workspace.setup();

        test_bed.widget.widget.init().unwrap();
        test_bed
            .widget
            .terminal
            .draw(|f| test_bed.widget.widget.draw(f, f.area()))
            .unwrap();

        assert_snapshot!(test_bed.widget.terminal.backend());

        test_bed.workspace.restore();
    }
}
