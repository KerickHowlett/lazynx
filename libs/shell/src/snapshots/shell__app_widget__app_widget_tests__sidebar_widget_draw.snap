---
source: libs/shell/src/app_widget.rs
expression: test_bed.widget.terminal.backend()
---
"╭─[1]─Workspace──────────────╮"
"│   test                     │"
"╰────────────────────────────╯"
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
"                              "
